nixos/modules/system/etc/build-composefs-dump.py

```diff
@@ -0,0 +1,198 @@
+#!/usr/bin/env python
+
+"""Build a composefs dump from a Json config"""
+
+import functools
+import glob
+import json
+import os
+import sys
+from enum import Enum
+from typing import Any
+
+Attrs = dict[str, Any]
+
+
+class FileType(Enum):
+    directory = "4"
+    file = "10"
+    symlink = "12"
+
+
+class ComposefsPath:
+    path: str
+    size: int
+    filetype: FileType
+    mode: str
+    uid: str
+    gid: str
+    payload: str
+    rdev: str = "0"
+    nlink: int = 1
+    mtime: str = "1.0"
+    content: str = "-"
+    digest: str = "-"
+
+    def __init__(
+        self,
+        attrs: Attrs,
+        size: int,
+        filetype: FileType,
+        mode: str,
+        payload: str,
+        path: str | None = None,
+    ):
+        if path is None:
+            path = attrs["target"]
+        self.path = "/" + path
+        self.size = size
+        self.filetype = filetype
+        self.mode = mode
+        self.uid = attrs["uid"]
+        self.gid = attrs["gid"]
+        self.payload = payload
+
+    def write_line(self) -> str:
+        line_list = [
+            str(self.path),
+            str(self.size),
+            f"{self.filetype.value}{self.mode}",
+            str(self.nlink),
+            str(self.uid),
+            str(self.gid),
+            str(self.rdev),
+            str(self.mtime),
+            str(self.payload),
+            str(self.content),
+            str(self.digest),
+        ]
+        return " ".join(line_list)
+
+
+def eprint(s: Any) -> None:
+    print(s, file=sys.stderr)
+
+
+def leading_directories(path: str) -> list[str]:
```

Comment:
Is this not the same as Path(path).parent?

Reply:
I thought it's the same as [Path(path).parents](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.parents).